janus_server/src/message.rs

```diff
@@ -0,0 +1,1378 @@
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct Interval {
+    /// The start of the interval.
+    pub start: Time,
+    /// The length of the interval.
+    pub duration: Duration,
+}
+
+impl Encode for Interval {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        self.start.encode(bytes);
+        self.duration.encode(bytes);
+    }
+}
+
+impl Decode for Interval {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let start = Time::decode(bytes)?;
+        let duration = Duration::decode(bytes)?;
+
+        Ok(Self { start, duration })
+    }
+}
+
+/// PPM protocol message representing a nonce uniquely identifying a client report.
+#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Nonce {
+    /// The time at which the report was generated.
+    pub time: Time,
+    /// A randomly generated value.
+    pub rand: u64,
+}
+
+impl Encode for Nonce {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        self.time.encode(bytes);
+        self.rand.encode(bytes);
+    }
+}
+
+impl Decode for Nonce {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let time = Time::decode(bytes)?;
+        let rand = u64::decode(bytes)?;
+
+        Ok(Self { time, rand })
+    }
+}
+
+/// PPM protocol message representing the different roles that participants can adopt.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, TryFromPrimitive)]
+#[repr(u8)]
+pub enum Role {
+    Collector = 0,
+    Client = 1,
+    Leader = 2,
+    Helper = 3,
+}
+
+impl Encode for Role {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        (*self as u8).encode(bytes);
+    }
+}
+
+impl Decode for Role {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let val = u8::decode(bytes)?;
+        Self::try_from(val)
+            .map_err(|_| CodecError::Other(anyhow!("unexpected Role value {}", val).into()))
+    }
+}
+
+/// PPM protocol message representing an identifier for an HPKE config.
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+pub struct HpkeConfigId(pub u8);
+
+impl Encode for HpkeConfigId {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        self.0.encode(bytes);
+    }
+}
+
+impl Decode for HpkeConfigId {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        Ok(Self(u8::decode(bytes)?))
+    }
+}
+
+/// PPM protocol message representing an HPKE ciphertext.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct HpkeCiphertext {
+    /// An identifier of the HPKE configuration used to seal the message.
+    pub config_id: HpkeConfigId,
+    /// An encasulated HPKE context.
+    pub encapsulated_context: Vec<u8>,
+    /// An HPKE ciphertext.
+    pub payload: Vec<u8>,
+}
+
+impl Encode for HpkeCiphertext {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        self.config_id.encode(bytes);
+        encode_u16_items(bytes, &(), &self.encapsulated_context);
+        encode_u16_items(bytes, &(), &self.payload);
```

Comment:
Ah, we should probably introduce variants of `encode_$SIZE_items` generic over `Encode` that omit the decoding parameter. That's not needed right here though.

Reply:
Indeed: I left those functions for later with the PR to split parameterized from non-parameterized encoding/decoding, but with this code change it seems like it would be useful.